validation/src/header.rs

```diff
@@ -270,52 +274,35 @@ fn compute_next_difficulty(
     let last_adjustment_time = last_adjustment_header.time;
 
     // Computing the time interval between the last adjustment header time and
-    // current time. The expected value actual_interval is 2 weeks assuming
+    // current time. The expected value timespan is 2 weeks assuming
     // the expected block time is 10 mins. But most of the time, the
-    // actual_interval will deviate slightly from 2 weeks. Our goal is to
+    // timespan will deviate slightly from 2 weeks. Our goal is to
     // readjust the difficulty target so that the expected time taken for the next
     // 2016 blocks is again 2 weeks.
     // IMPORTANT: The bitcoin protocol allows for a roughly 3-hour window around
     // timestamp (1 hour in the past, 2 hours in the future) meaning that
-    // the actual_interval can be negative on testnet networks.
-    let actual_interval = (prev_header.time as i64) - (last_adjustment_time as i64);
-
-    // The target_adjustment_interval_time is 2 weeks of time expressed in seconds
-    let target_adjustment_interval_time = (DIFFICULTY_ADJUSTMENT_INTERVAL * TEN_MINUTES) as i64;
-
-    // Adjusting the actual_interval to [0.5 week, 8 week] range in case the
-    // actual_interval deviates too much from the expected 2 weeks.
-    let adjusted_interval = actual_interval.clamp(
-        target_adjustment_interval_time / 4,
-        target_adjustment_interval_time * 4,
-    ) as u32;
-
-    // Computing new difficulty target.
-    // new difficulty target = old difficult target * (adjusted_interval /
-    // 2_weeks);
-    let mut target = prev_header.target();
-    target = target.mul_u32(adjusted_interval);
-    target = target / Uint256::from_u64(target_adjustment_interval_time as u64).unwrap();
-
-    // Adjusting the newly computed difficulty target so that it doesn't exceed the
-    // max_difficulty_target limit
-    target = Uint256::min(target, max_target(network));
-
-    // Converting the target (Uint256) into a 32 bit representation used by Bitcoin
-    Header::compact_target_from_u256(&target)
+    // the timespan can be negative on testnet networks.
+    let timespan = prev_header.time.saturating_sub(last_adjustment_time) as u64;
+
+    CompactTarget::from_next_work_required(prev_header.bits, timespan, Params::new(*network))
```

Comment:
Here's link to implementation details:
https://docs.rs/bitcoin/0.32.4/src/bitcoin/pow.rs.html#390-414

Reply:
It looks ok to me but I am also not the most familiar with the difficulty calculation anyway. Perhaps @THLO can also confirm here? We do have tests that I trust would have let us know if something was horribly off but an extra pair of eyes shouldn't hurt.

Reply:
It looks good to me.
We are now using the `from_next_work_required` function from the `bitcoin` crate, right? This function should be implemented correctly.

Reply:
> We are now using the from_next_work_required function from the bitcoin crate, right?

yes. I'm closing this thread due to LGTMs.